### ISSUE TITLE
False positive error with method override and generic type annotations in class decorator

### ISSUE THREAD

**Issue by gen740** [microsoft/pyright#9359]
I created the following wrapper for a task class, and Pyright emits a false positive error.

### Environment
- Python version: 3.12.5
- Pyright version: 1.1.382

### Code Example
```python
from typing import Callable


class BaseTask[INPUT, OUTPUT]:
    def task(self, input: INPUT) -> OUTPUT:
        raise NotImplementedError


def task_wrapper[INPUT, OUTPUT](
    task_input: type[INPUT],
    task_output: type[OUTPUT],
):
    def decorator(fun: Callable[[INPUT], OUTPUT]) -> type[BaseTask[INPUT, OUTPUT]]:
        class _Task(BaseTask[INPUT, OUTPUT]):
            def task(self, input: INPUT) -> OUTPUT:
                if isinstance(input, task_input):
                    raise ValueError(
                        f"Invalid input type: {type(input)} expected {task_input}"
                    )
                output = fun(input)
                if isinstance(output, task_output):
                    raise ValueError(
                        f"Invalid output type: {type(output)} expected {task_output}"
                    )
                return output

        return _Task

    return decorator
```

### Expected Behavior
The code should type-check without errors, as the return type of the overridden `task` method in `_Task` should be compatible with the original definition in `BaseTask`.

### Actual Behavior
Pyright emits an error:
```
$ pyright test.py
/Users/home/test.py
  /Users/home/test.py:15:17 - error: Method "task" overrides class "BaseTask" in an incompatible manner
    Return type mismatch: base method returns type "OUTPUT@task_wrapper", override returns type "OUTPUT@task_wrapper"
      Type "OUTPUT@task_wrapper" is not assignable to type "OUTPUT@task_wrapper" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations
```
Pyright reports an error, while mypy does not emit any errors for the same code.

**Comment by erictraut** [comment]
Thanks for the bug report and the clear repro example.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.388.
